File: apps/visual-oscar/lib/sources/ted.ts
/**
 * Adaptador para Tenders Electronic Daily (TED) — diario oficial de la UE
 * para licitaciones públicas europeas.
 *
 * API v3: https://api.ted.europa.eu/v3/notices/search (POST)
 * Filtramos a buyer-country=ESP para licitaciones publicadas por organismos
 * españoles en TED.
 */

import type { NormalizedContrato } from '@/lib/socrata-catalunya'

const ENDPOINT = 'https://api.ted.europa.eu/v3/notices/search'
const UA = 'Politeia-Analitica/1.0 (+https://politeia-analitica.vercel.app)'

interface TedNotice {
  'publication-number'?: string
  'title-of-procedure'?: { spa?: string; eng?: string; [k: string]: unknown }
  'buyer-name'?: Array<{ spa?: string; eng?: string; [k: string]: unknown }>
  'buyer-country'?: string[]
  'buyer-city'?: Array<{ spa?: string; eng?: string }>
  'contract-nature'?: string[]
  'classification-cpv'?: string[]
  'total-value'?: number | string
  'total-value-cur'?: string
  'publication-date'?: string
  'deadline-date-receipt-tender'?: string
  links?: { xml?: { MUL?: string }; pdf?: { ENG?: string; SPA?: string } }
}

function pickLang(o: unknown, langs = ['spa', 'eng']): string | undefined {
  if (!o) return undefined
  if (typeof o === 'string') return o
  if (Array.isArray(o)) return pickLang(o[0])
  if (typeof o === 'object') {
    for (const l of langs) {
      const v = (o as Record<string, unknown>)[l]
      if (typeof v === 'string') return v
    }
    const vals = Object.values(o as Record<string, unknown>)
    for (const v of vals) if (typeof v === 'string') return v
  }
  return undefined
}

function normalize(n: TedNotice): NormalizedContrato {
  const titulo = pickLang(n['title-of-procedure'])
  const buyer = pickLang(n['buyer-name'])
  const cpv = Array.isArray(n['classification-cpv']) ? n['classification-cpv'][0] : undefined
  const fecha = n['publication-date']?.slice(0, 10)
  const importeRaw = n['total-value']
  const importe = typeof importeRaw === 'number' ? importeRaw
    : typeof importeRaw === 'string' ? Number(importeRaw) : undefined
  return {
    id: `TED-${n['publication-number']}`,
    source_id: n['publication-number'],
    fuente: 'CATALUNYA_SOCRATA' as never,
    fuente_label: 'TED · Diario UE',
    expediente: n['publication-number'] || '',
    organo: buyer || '—',
    ambito: 'Europa',
    objeto: titulo || '—',
    tipo_contrato: Array.isArray(n['contract-nature']) ? n['contract-nature'][0] : undefined,
    cpv,
    cpv_div: cpv?.slice(0, 2),
    lugar_ejecucion: pickLang(n['buyer-city']),
    importe_licitacion: undefined,
    importe_adjudicacion: Number.isFinite(importe) ? importe : undefined,
    estado: 'Publicado',
    fecha_publicacion: fecha,
    anio: fecha ? Number(fecha.slice(0, 4)) : undefined,
    url: n.links?.pdf?.SPA || n.links?.pdf?.ENG || (n['publication-number'] ? `https://ted.europa.eu/es/notice/${n['publication-number']}` : undefined),
  }
}

export interface TedSearchParams {
  q?: string
  desde?: string                   // ISO YYYY-MM-DD
  hasta?: string
  cpv_div?: string
  organo?: string
  limit?: number
  page?: number
}

/**
 * Construye la query SoQL-like de TED. Filtramos a España siempre.
 *
 * Sintaxis TED v3:
 *   campo="valor"          (exacto)
 *   campo>=fecha           (rango)
 *   FT="texto"             (full text)
 *   AND/OR/NOT             (booleanos)
 *
 * Docs: https://docs.ted.europa.eu/api/index.html
 */
function buildQuery(p: TedSearchParams): string {
  const parts: string[] = ['buyer-country="ESP"']
  if (p.desde) parts.push(`publication-date>=${p.desde}`)
  if (p.hasta) parts.push(`publication-date<=${p.hasta}`)
  if (p.cpv_div) parts.push(`classification-cpv=${p.cpv_div}*`)
  if (p.q) parts.push(`FT="${p.q.replace(/"/g, '\\"')}"`)
  if (p.organo) parts.push(`buyer-name="${p.organo.replace(/"/g, '\\"')}"~`)
  return parts.join(' AND ')
}

export async function searchTed(
  p: TedSearchParams,
  timeoutMs = 8000,
): Promise<{ ok: boolean; items: NormalizedContrato[]; ms: number; error?: string }> {
  const t0 = Date.now()
  const limit = Math.min(100, Math.max(1, p.limit ?? 30))
  const page = Math.max(1, p.page ?? 1)

  const body = {
    query: buildQuery(p),
    fields: [
      'publication-number',
      'title-of-procedure',
      'buyer-name',
      'buyer-country',
      'buyer-city',
      'contract-nature',
      'classification-cpv',
      'total-value',
      'total-value-cur',
      'publication-date',
      'deadline-date-receipt-tender',
      'links',
    ],
    page,
    limit,
  }

  try {
    const ctrl = new AbortController()
    const timer = setTimeout(() => ctrl.abort(), timeoutMs)
    const res = await fetch(ENDPOINT, {
      method: 'POST',
      headers: {
        'User-Agent': UA,
        'Content-Type': 'application/json',
        Accept: 'application/json',
      },
      body: JSON.stringify(body),
      signal: ctrl.signal,
      next: { revalidate: 600 },
    })
    clearTimeout(timer)
    if (!res.ok) return { ok: false, items: [], ms: Date.now() - t0, error: `HTTP ${res.status}` }
    const json = (await res.json()) as { notices?: TedNotice[]; totalNoticeCount?: number }
    const items = (json.notices || []).map(normalize).map(n => ({
      ...n,
      fuente: 'TED' as 'CATALUNYA_SOCRATA',
    }))
    return { ok: true, items, ms: Date.now() - t0 }
  } catch (e: unknown) {
    return { ok: false, items: [], ms: Date.now() - t0, error: e instanceof Error ? e.message : 'unknown' }
  }
}

File: etl/sources/eu/ted.py
"""Conector TED (Tenders Electronic Daily · UE) · Sprint 3 · S3.2.

> **Sprint 3 · S3.2** (`docs/ROADMAP_GITS_AMIGOS.md §4 Sprint 3`)

TED (Tenders Electronic Daily) es el portal oficial de licitaciones públicas
de la UE. Contiene **TODAS** las licitaciones por encima de umbrales UE
(~130 000€ servicios, ~5M€ obras) en 30+ países desde 1993.

API REST oficial: https://api.ted.europa.eu/v3
Documentación: https://docs.ted.europa.eu/api/

Endpoints principales:
  - POST /notices/search          → busqueda por filtros (fields, query)
  - GET  /notices/{publication-number}  → notice individual

Filtros relevantes para Politeia:
  - place-of-performance: ESP (solo España)
  - notice-type: PIN, CN, CAN, MOD, ...
  - publication-date: YYYY-MM-DD..YYYY-MM-DD
  - classification-cpv: códigos CPV de sector
  - estimated-total-value: importe mínimo

Output: produce NormalizedItem para cada licitacion encontrada.

Falla cerrado: si API caída o rate limit, devuelve [] sin romper.
"""
from __future__ import annotations

import hashlib
import logging
from datetime import datetime, timezone
from typing import Any, Iterator

logger = logging.getLogger(__name__)


_BASE_URL = "https://api.ted.europa.eu/v3"
_TIMEOUT = 25
_USER_AGENT = "Politeia-Analitica/2.0 (+https://politeia-analitica.es)"


# Códigos CPV (Common Procurement Vocabulary) de sectores Politeia
# Lista parcial · expandir cuando se añadan más sectores
CPV_BY_SECTOR: dict[str, list[str]] = {
    "energia":           ["09000000", "31000000", "65000000"],   # productos energéticos · infraestructura energética
    "farma":             ["33000000", "85000000"],                # productos médicos · servicios salud
    "defensa":           ["35000000"],                            # defensa
    "infraestructuras":  ["45000000", "71000000"],                # obras + ingeniería
    "telecom":           ["64000000", "72000000"],                # telecom + servicios IT
    "transporte":        ["60000000", "63000000"],                # transporte + servicios apoyo
    "agroalimentario":   ["15000000", "03000000"],                # alimentación + agricultura
    "vivienda":          ["70000000"],                            # inmobiliario
    "educacion":         ["80000000"],                            # educación
}


# ────────────────────────────────────────────────────────────────────
# Cliente TED · falla cerrado
# ────────────────────────────────────────────────────────────────────

class TEDClient:
    """Cliente para la API REST de TED EU.

    Nunca lanza excepción hacia arriba: retorna None / [] en caso de error.
    """

    def __init__(self, session: Any = None) -> None:
        try:
            import requests  # type: ignore
            self._session = session or requests.Session()
            self._session.headers.update({
                "Accept": "application/json",
                "Content-Type": "application/json",
                "User-Agent": _USER_AGENT,
            })
        except ImportError:
            self._session = None
            logger.warning("TEDClient: requests no disponible · degradado")

    def search_notices(
        self,
        *,
        country: str = "ESP",
        date_from: str | None = None,
        date_to: str | None = None,
        cpv_codes: list[str] | None = None,
        query_text: str | None = None,
        notice_types: list[str] | None = None,
        page: int = 1,
        page_size: int = 50,
    ) -> dict[str, Any] | None:
        """Busca notices (licitaciones) por filtros combinados.

        Args:
          country: ISO 3166-1 alpha-3 (ESP, FRA, DEU, ...) · default ESP
          date_from / date_to: 'YYYYMMDD' (formato TED) o 'YYYY-MM-DD'
          cpv_codes: lista de códigos CPV
          query_text: búsqueda en texto libre
          notice_types: ['PIN', 'CN', 'CAN', 'MOD']
          page: 1-based
          page_size: max 250 según API

        Returns:
          dict con {total, notices: [...], error}  o None si fallo grave
        """
        # Construir query string en lenguaje de TED Expert Search
        clauses: list[str] = []
        if country:
            clauses.append(f"place-of-performance={country}")
        if date_from and date_to:
            df = date_from.replace("-", "")
            dt = date_to.replace("-", "")
            clauses.append(f"publication-date>={df} AND publication-date<={dt}")
        elif date_from:
            df = date_from.replace("-", "")
            clauses.append(f"publication-date>={df}")
        if cpv_codes:
            cpv_clause = " OR ".join(f"classification-cpv={c}" for c in cpv_codes)
            clauses.append(f"({cpv_clause})")
        if notice_types:
            nt_clause = " OR ".join(f"notice-type={t}" for t in notice_types)
            clauses.append(f"({nt_clause})")

        query = " AND ".join(clauses) if clauses else "*"
        if query_text:
            query = f"({query}) AND \"{query_text}\""

        payload = {
            "query": query,
            "fields": [
                "publication-number",
                "notice-title",
                "buyer-name",
                "place-of-performance",
                "publication-date",
                "deadline-date-lots",
                "classification-cpv",
                "estimated-total-value",
                "links",
            ],
            "page": page,
            "limit": min(page_size, 250),
            "scope": "ACTIVE",  # solo notices vigentes
        }

        return self._post_json(f"{_BASE_URL}/notices/search", payload)

    def get_notice(self, publication_number: str) -> dict[str, Any] | None:
        """Devuelve un notice por número de publicación (ej '12345-2026')."""
        if not publication_number:
            return None
        return self._get_json(f"{_BASE_URL}/notices/{publication_number}")

    # ── HTTP helpers ──────────────────────────────────────────────

    def _post_json(self, url: str, json_body: dict) -> dict | None:
        if self._session is None:
            return None
        try:
            r = self._session.post(url, json=json_body, timeout=_TIMEOUT)
            if r.status_code == 404:
                return None
            r.raise_for_status()
            return r.json()
        except Exception as exc:
            logger.warning("TED POST error · %s · %s", url[:80], exc)
            return None

    def _get_json(self, url: str, params: dict | None = None) -> dict | None:
        if self._session is None:
            return None
        try:
            r = self._session.get(url, params=params, timeout=_TIMEOUT)
            if r.status_code == 404:
                return None
            r.raise_for_status()
            return r.json()
        except Exception as exc:
            logger.warning("TED GET error · %s · %s", url[:80], exc)
            return None


# ────────────────────────────────────────────────────────────────────
# Adapter · TED notice → NormalizedItem
# ────────────────────────────────────────────────────────────────────

def _notice_to_normalized(raw: dict[str, Any]) -> Any | None:
    """Convierte un TED notice raw → NormalizedItem.

    Mapeo de campos TED → NormalizedItem:
      raw.publication-number → item_id
      raw.notice-title.spa o .eng → title
      raw.buyer-name → author
      raw.publication-date → published_at
      raw.links.html.spa → url
      raw.estimated-total-value → payload.importe_estimado
      raw.classification-cpv → payload.cpv_codes
      raw.place-of-performance → payload.lugar_ejecucion
    """
    try:
        from packages.types import NormalizedItem
    except ImportError:
        return None

    pub_num = raw.get("publication-number") or raw.get("publicationNumber") or ""
    if not pub_num:
        return None

    # Title · puede ser dict multilingüe {spa: '...', eng: '...'}
    title_obj = raw.get("notice-title") or raw.get("noticeTitle") or {}
    if isinstance(title_obj, dict):
        title = title_obj.get("spa") or title_obj.get("eng") or next(iter(title_obj.values()), "")
    else:
        title = str(title_obj)

    if not title:
        return None

    # Fecha
    fecha_str = raw.get("publication-date") or raw.get("publicationDate") or ""
    try:
        if fecha_str:
            # TED usa YYYYMMDD o ISO
            if len(fecha_str) == 8 and fecha_str.isdigit():
                published_at = datetime(int(fecha_str[:4]), int(fecha_str[4:6]), int(fecha_str[6:8]), tzinfo=timezone.utc)
            else:
                published_at = datetime.fromisoformat(fecha_str.replace("Z", "+00:00"))
        else:
            published_at = datetime.now(timezone.utc)
    except Exception:
        published_at = datetime.now(timezone.utc)

    # URL
    links = raw.get("links") or {}
    url_html = ""
    if isinstance(links, dict):
        html_links = links.get("html", {})
        if isinstance(html_links, dict):
            url_html = html_links.get("spa") or html_links.get("eng") or next(iter(html_links.values()), "")
        elif isinstance(html_links, list) and html_links:
            url_html = str(html_links[0])

    # Buyer
    buyer = raw.get("buyer-name") or raw.get("buyerName") or ""
    if isinstance(buyer, dict):
        buyer = buyer.get("spa") or buyer.get("eng") or next(iter(buyer.values()), "")
    elif isinstance(buyer, list) and buyer:
        buyer = str(buyer[0])

    # Hash
    raw_hash = hashlib.sha256(f"ted|{pub_num}|{title}".encode("utf-8")).hexdigest()

    payload: dict[str, Any] = {
        "publication_number": pub_num,
        "buyer": str(buyer)[:240],
        "lugar_ejecucion": raw.get("place-of-performance") or raw.get("placeOfPerformance") or "",
        "cpv_codes": raw.get("classification-cpv") or raw.get("classificationCpv") or [],
        "deadline": raw.get("deadline-date-lots") or raw.get("deadlineDate") or "",
        "importe_estimado": raw.get("estimated-total-value") or raw.get("estimatedTotalValue") or {},
    }

    try:
        return NormalizedItem(
            source="ted_eu",
            item_id=str(pub_num),
            title=str(title)[:2000].strip(),
            body="",  # TED API no devuelve cuerpo · queda para una segunda llamada get_notice
            summary=str(title)[:400].strip(),
            url=url_html or None,
            published_at=published_at,
            author=str(buyer)[:240],
            language="es",
            raw_hash=raw_hash,
            categories=["licitacion", "ted_eu"],
            payload=payload,
        )
    except Exception as exc:
        logger.debug("TED NormalizedItem build · %s", exc)
        return None


# ────────────────────────────────────────────────────────────────────
# Generador NormalizedItem
# ────────────────────────────────────────────────────────────────────

def to_normalized_items(
    *,
    country: str = "ESP",
    sector: str | None = None,
    date_from: str | None = None,
    date_to: str | None = None,
    max_items: int = 100,
) -> Iterator[Any]:
    """Genera NormalizedItem desde TED para licitaciones recientes.

    Args:
      country: ESP por default · ISO alpha-3
      sector: clave de CPV_BY_SECTOR para auto-filtrar por sectores Politeia
      date_from / date_to: 'YYYY-MM-DD'
      max_items: máximo a generar
    """
    client = TEDClient()
    if client._session is None:
        return

    cpv_codes = CPV_BY_SECTOR.get(sector, []) if sector else None

    yielded = 0
    page = 1
    while yielded < max_items:
        result = client.search_notices(
            country=country,
            date_from=date_from,
            date_to=date_to,
            cpv_codes=cpv_codes,
            page=page,
            page_size=50,
        )
        if not result:
            break

        notices = result.get("notices") or result.get("results") or []
        if not notices:
            break

        for raw in notices:
            if yielded >= max_items:
                break
            norm = _notice_to_normalized(raw)
            if norm:
                yield norm
                yielded += 1

        if len(notices) < 50:
            break
        page += 1


# ────────────────────────────────────────────────────────────────────
# Singleton
# ────────────────────────────────────────────────────────────────────

_CLIENT: TEDClient | None = None


def get_ted_client() -> TEDClient:
    global _CLIENT
    if _CLIENT is None:
        _CLIENT = TEDClient()
    return _CLIENT


__all__ = [
    "TEDClient",
    "CPV_BY_SECTOR",
    "get_ted_client",
    "to_normalized_items",
]
